services/repository/files/tree.go

```diff
@@ -118,3 +122,389 @@ func GetTreeBySHA(ctx context.Context, repo *repo_model.Repository, gitRepo *git
+		}
+		return nodes[i].Name < nodes[j].Name
+	})
+}
+
+/*
+Example 1: (path: /)
+
+	GET /repo/name/tree/
+
+	resp:
+	[{
+	    "name": "d1",
+	    "isFile": false,
+	    "path": "d1"
+	},{
+	    "name": "d2",
+	    "isFile": false,
+	    "path": "d2"
+	},{
+	    "name": "d3",
+	    "isFile": false,
+	    "path": "d3"
+	},{
+	    "name": "f1",
+	    "isFile": true,
+	    "path": "f1"
+	},]
+
+Example 2: (path: d3)
+
+	GET /repo/name/tree/d3
+	resp:
+	[{
+	    "name": "d3d1",
+	    "isFile": false,
+	    "path": "d3/d3d1"
+	}]
+
+Example 3: (path: d3/d3d1)
+
+	GET /repo/name/tree/d3/d3d1
+	resp:
+	[{
+	    "name": "d3d1f1",
+	    "isFile": true,
+	    "path": "d3/d3d1/d3d1f1"
+	},{
+	    "name": "d3d1f1",
+	    "isFile": true,
+	    "path": "d3/d3d1/d3d1f2"
+	}]
+*/
+func GetTreeList(ctx context.Context, repo *repo_model.Repository, gitRepo *git.Repository, treePath string, ref git.RefName, recursive bool) ([]*TreeViewNode, error) {
+	if repo.IsEmpty {
+		return nil, nil
+	}
+	if ref == "" {
+		ref = git.RefNameFromBranch(repo.DefaultBranch)
+	}
+
+	// Check that the path given in opts.treePath is valid (not a git path)
+	cleanTreePath := CleanUploadFileName(treePath)
+	if cleanTreePath == "" && treePath != "" {
+		return nil, ErrFilenameInvalid{
+			Path: treePath,
+		}
+	}
+	treePath = cleanTreePath
+
+	// Get the commit object for the ref
+	commit, err := gitRepo.GetCommit(ref.String())
+	if err != nil {
+		return nil, err
+	}
+
+	entry, err := commit.GetTreeEntryByPath(treePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the entry is a file, we return a FileContentResponse object
+	if entry.Type() != "tree" {
+		return nil, fmt.Errorf("%s is not a tree", treePath)
+	}
+
+	gitTree, err := commit.SubTree(treePath)
+	if err != nil {
+		return nil, err
+	}
+	var entries git.Entries
+	if recursive {
+		entries, err = gitTree.ListEntriesRecursiveFast()
+	} else {
+		entries, err = gitTree.ListEntries()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var treeViewNodes []*TreeViewNode
+	mapTree := make(map[string][]*TreeViewNode)
+	for _, e := range entries {
+		subTreePath := path.Join(treePath, e.Name())
+
+		if strings.Contains(e.Name(), "/") {
+			mapTree[path.Dir(e.Name())] = append(mapTree[path.Dir(e.Name())], &TreeViewNode{
+				Name: path.Base(e.Name()),
+				Type: entryModeString(e.Mode()),
+				Path: subTreePath,
+			})
+		} else {
+			treeViewNodes = append(treeViewNodes, &TreeViewNode{
+				Name: e.Name(),
+				Type: entryModeString(e.Mode()),
+				Path: subTreePath,
+			})
+		}
+	}
+
+	for _, node := range treeViewNodes {
+		if node.Type == "tree" {
+			node.Children = mapTree[node.Path]
+			sortTreeViewNodes(node.Children)
+		}
+	}
+
+	sortTreeViewNodes(treeViewNodes)
+
+	return treeViewNodes, nil
+}
+
+// GetTreeInformation returns the first level directories and files and all the trees of the path to treePath.
+// If treePath is a directory, list all subdirectories and files of treePath.
+/*
+Example 1: (path: /)
```

Comment:
What's the difference from `GetTreeList`? The code is almost the same and the examples are copied?

Reply:
`GetTreeList` is only used when expanding a single folder and does not support the expansion of multiple layers of folders.
`GetTreeInformation` is used when expanding multiple layers of folders. For the specific differences, you can refer to the example in the comments.

Reply:
I do not see why it is necessary or how it could be right.

Rewritten.